### ISSUE TITLE
huge webpack-stats produce a JavaScript heap out of memory

### ISSUE THREAD

**Issue by batusai513** [zeit/next.js#3404]
We are working on a Very large project with several dependencies, when building the project we get a  `FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory`.

- [ x] I have searched the [issues](https://github.com/zeit/next.js/issues) of this repository and believe that this is not a duplicate.

## Expected Behavior

When running `yarn build` I should get a working production bundle

## Current Behavior

When running `yarn build` I get `FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory`.

After several hours of debugging I found out that the origin of this problem comes from https://github.com/zeit/next.js/blob/canary/server/build/index.js#L72
more exactly the `JSON.stringify` function, given that our stats file is 260MB, node can't handle it.

I was able to patch it using streams and make it work ok, I'll make a PR for this tomorrow.

## Your Environment

| Tech    | Version |
|---------|---------|
| next    |    4.1.4     |
| node    |    8.9.1     |
| OS      |     Ubuntu 16.04    |
| browser |    N/A     |

**Comment by webernir** [comment]
@batusai513  seems like #3150 have caused it:
https://github.com/zeit/next.js/pull/3150/files

**Comment by ShepelievD** [comment]
@batusai513 faced the same with 4.1.4 version

what I did with my colleague:

```
const UglifyJSPlugin = require('uglifyjs-webpack-plugin')

module.exports = {
  webpack: (config, { dev }) => {
    if (!dev) {
      for (let index = 0; index < config.plugins.length; index += 1) {
        if (config.plugins[index].constructor.name === 'UglifyJsPlugin') {
          config.plugins.splice(index, 1, new UglifyJSPlugin({
            sourceMap: false,
            parallel: true
          }))
          break
        }
      }
    }

    // Perform customizations to config
    if (dev) {
      config.module.rules.push({
        test: /\.js$/,
        exclude: /node_modules/,
        loader: 'eslint-loader',
        options: {
        }
      })
    }

    if (config.resolve.alias) {
      delete config.resolve.alias['react']
      delete config.resolve.alias['react-dom']
    }

    return config
  }
}
```

The same is not working with v4.2.0

**Comment by ShepelievD** [comment]
@webernir looks like it is 4.2.0 version, but issue still exists there, they do what we did above

**Comment by batusai513** [comment]
@webernir UglifyJS is not the problem causing the heap memory error, yes, it takes a long time with the current configuration but the root cause for the memory error is the JSON.stringify described before.

@ShepelievD yeap, i did the same and got more bearable building time, about six minutes.

**Comment by webernir** [comment]
@batusai513 Can you share a reference for the JSON.stringify problem?

**Comment by batusai513** [comment]
@webernir https://github.com/nodejs/node/issues/13465

**Comment by timneutkens** [comment]
Fixed in 4.2.1
